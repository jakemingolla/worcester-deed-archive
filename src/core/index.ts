export * from "./log";
export * from "./screenshot";
export * from "./config";
