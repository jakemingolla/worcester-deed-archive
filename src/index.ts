import { stringify } from "csv-stringify/sync";

import { config, log, screenshot } from "./core/index.ts";
import { countResults } from "./steps/count-results.ts";
import { getRow } from "./steps/get-row.ts";
import { goToNextPage } from "./steps/go-to-next-page.ts";
import { init } from "./steps/init.ts";
import { setSearchCriteria } from "./steps/set-search-criteria.ts";
import { validateTimeRange } from "./steps/validate-time-range.ts";

export const main = async (
  startInput: string,
  endInput: string,
): Promise<string> => {
  const { start, end } = validateTimeRange(startInput, endInput);
  log.info(
    `Beginning run at ${
      config.BASE_URL
    } from ${start.format()} -> ${end.format()}.`,
  );

  const { browser, page } = await init(1);

  try {
    await page.goto(config.BASE_URL);

    await page.waitForFunction(() => typeof __doPostBack !== "undefined");
    await screenshot.record(page, "after-do-post-back");

    await setSearchCriteria(page, start, end);
    await screenshot.record(page, "after-search-criteria");
    await page.click("#SearchFormEx1_btnSearch");

    const count = await countResults(page);
    await screenshot.record(page, "after-count");
    log.info(`Successfully found ${count} results.`);

    const headers = [
      [
        // General details
        "Doc #",
        "Rec Date",
        "Rec Time",
        "Type Desc.",
        "# of Pages",
        "Book/Page",
        "Consideration",
        "Court Case #",
        // Property details
        "Street #",
        "Street Name",
        "Description",
        // TODO grantees?
      ],
    ];
    const data = [];

    // TODO config page size
    for (let i = 0; i < count; i++) {
      const fn = `DocList1$GridView_Document$ctl${((i % 20) + 2)
        .toString()
        .padStart(2, "0")}$ButtonRow_Rec. Date_${i % 20}`;
      await page.evaluate((fn) => {
        return __doPostBack(fn, "");
      }, fn);

      const id = `DocList1_GridView_Document_ctl${((i % 20) + 2)
        .toString()
        .padStart(2, "0")}_ButtonRow_Rec. Date_${i % 20}`;
      await page.waitForXPath(
        `//*[@id="${id}"]/ancestor::tr[@class="DataGridSelectedRow"]`,
      );

      data.push(await getRow(page));
      await Bun.sleep(config.AFTER_EACH_ROW_DELAY_SECONDS * 1000);

      if (i % 20 === 0 || i % 20 === 19) {
        await screenshot.record(page, `after-${i}`);
      }

      if (i % 20 === 19 && i + 1 < count) {
        await goToNextPage(page, i);
        await Bun.sleep(config.AFTER_EACH_PAGE_DELAY_SECONDS * 1000);
      }
      log.debug(`Successfully processed i = ${i} of ${count} results.`);
    }

    return stringify(headers.concat(data));
  } catch (err) {
    console.log(err);
    await screenshot.record(page, "error-received");
    throw err;
  } finally {
    await browser.close();
  }
};

if (import.meta.main) {
  main(config.START_DATE, config.END_DATE)
    .then((output) => {
      console.log(output);
      return Bun.write(config.OUTPUT_FILE, output);
    })
    .then(() => process.exit(0))
    .catch((err) => {
      log.error((err as Error).stack);
      return process.exit(1);
    });
}
