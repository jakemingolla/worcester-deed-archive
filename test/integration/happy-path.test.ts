import { describe, it, expect } from "bun:test";
import { main } from "../../src";
import dayjs from "dayjs";

describe("happy path tests", () => {
  it("the results from 2023-01-01 -> 2023-02-01 are found", async () => {
    const start = dayjs("2023-01-01");
    const end = dayjs("2023-02-01");
    const output = await main(start, end);
    const expected =
      "Doc #,File Date,Rec Time,Type Desc.,Bk/Pg,Consideration,Court Case #,Doc. Status\n" +
      "18527,01/19/2023,08:46:19.602,PLANS,00093/127,\xa0,0000019658,In workflow\n";

    expect(output).toBe(expected);
  });
});
