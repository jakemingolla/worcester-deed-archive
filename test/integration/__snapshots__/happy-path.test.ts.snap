// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`happy path tests the results from 2023-01-01 -> 2023-02-01 are found 1`] = `
"Doc #,Rec Date,Rec Time,Type Desc.,# of Pages,Book/Page,Consideration,Court Case #,Street #,Street Name,Description,Town,Grantor,Grantee
416826,01/06/1961,00:00AM,DEED,1,04168/36,,Verified/Certified,,,,CHARLTON,PETERSON ROBERT F,DUNDERDALE CHARLES D
416827,01/06/1961,00:00AM,DEED,2,04168/37,100.00,Verified/Certified,,OLD DUDLEY RD,,OXFORD,CORTIS NORMAN E,JAVERY FRANK R
416830,01/06/1961,00:00AM,DEED,2,04168/40,100.00,Verified/Certified,,,,NORTHBRIDGE,KURTYKE JOHN L JR,KURTYKE DORIS G
416833,01/06/1961,00:00AM,DEED,2,04168/43,,Verified/Certified,,,,UXBRIDGE,HOLT CHARLES J,RICHARDSON DANIEL A
416844,01/06/1961,00:00AM,DEED,1,04168/55,100.00,Verified/Certified,,BEAVER ST,,WORCESTER,MORTON FLORENCE G,DAMELIO ALBERT V
416854,01/06/1961,00:00AM,DEED,1,04168/64,,Verified/Certified,,LAKE VIEW DR,,EAST BROOKFIELD,STEPHAN DISTRIBUTING CORP,STEPHAN CO
416855,01/06/1961,00:00AM,DEED,2,04168/65,,Verified/Certified,,CRANBROOK DR,,HOLDEN,ROBERT F DESMARAIS CONSTRUCTION INC,ADAMS GEORGE C
416865,01/06/1961,00:00AM,DEED,3,04168/77,,Verified/Certified,,COOPER RD,,NORTHBRIDGE,WRIGHT ROY L,BERGERON HOWARD
416868,01/06/1961,00:00AM,DEED,2,04168/80,,Verified/Certified,,WALNUT ST,,BERLIN,LAWRENCE J RICHARD,COLDWELL ROBERT B
416869,01/06/1961,00:00AM,DEED,2,04168/82,,Verified/Certified,,SOUTH MAIN ST,,DUDLEY,SZCZEPANIAK JOHN B,STEFANIK FRANK P
416870,01/06/1961,00:00AM,DEED,2,04168/83,,Verified/Certified,,YUKON ST,,WORCESTER,HOLLOWAY JACK,HUTTON GRAHAM M
416871,01/06/1961,00:00AM,DEED,2,04168/85,7500.00,Verified/Certified,181,MAIN ST,,BOYLSTON,WORCESTER COUNTY COMMRS,BILAZARIAN AGAVNY
416872,01/06/1961,00:00AM,DEED,2,04168/86,,Verified/Certified,,MARTIN ST,,MILLBURY,VALENTINO ANTHONY,MILLBURY DIE CO INC
416874,01/06/1961,00:00AM,DEED,2,04168/90,,Verified/Certified,,WEST MAIN ST,,UPTON,MANGO CARMINO,2
416875,01/06/1961,00:00AM,DEED,2,04168/91,,Verified/Certified,,,,SHREWSBURY,PERRON RALPH J,GRAVES RUSSELL D
416877,01/06/1961,00:00AM,DEED,2,04168/94,,Verified/Certified,,SUMMER ST,,NORTH BROOKFIELD,GAULT SARAH L,SEARS LEONARD F
416879,01/06/1961,00:00AM,DEED,2,04168/98,,Verified/Certified,,HOPKINTON RD,,SOUTHBORO,DEMONE JAMES A,NARICE ARTHUR
416881,01/06/1961,00:00AM,DEED,1,04168/100,,Verified/Certified,,,,MILLBURY,GENTILE NICHOLAS NOEL,JOHNSON STANLEY E
416884,01/06/1961,00:00AM,DEED,1,04168/104,100.00,Verified/Certified,,DANIELS ST,,HOPEDALE,YOUNG BARBARA E,YOUNG BARBARA E
416891,01/06/1961,00:00AM,DEED,2,04168/111,,Verified/Certified,,SOUTH QUINSIGAMOND AVE,,SHREWSBURY,ADAMS JOHN J,ADAMS RICHARD H
416893,01/06/1961,00:00AM,DEED,1,04168/113,500.00,Verified/Certified,,ROSEMONT RD,,WORCESTER,STARLING CLARENCE G,B A SUNDIN & SON INC
416894,01/06/1961,00:00AM,DEED,1,04168/114,,Verified/Certified,,GLEN ELLEN RD,,WORCESTER,HALLEN RUTH E,BROWN JULIUS J
416895,01/06/1961,00:00AM,DEED,2,04168/115,,Verified/Certified,,ROSEMONT RD,,WORCESTER,STARLING CLARENCE G,B A SUNDIN & SON INC
4168108,01/06/1961,00:00AM,DEED,2,04168/133,,Verified/Certified,,MILL ST,,LANCASTER,HUTCHINS WILLIAM A,HUTCHINS DONALD W
4168111,01/06/1961,00:00AM,DEED,1,04168/138,,Verified/Certified,,ANCHOR ST,,WORCESTER,CARRIGAN EVA C,MARTINEAU CHARLES E
4168114,01/06/1961,00:00AM,DEED,1,04168/142,,Verified/Certified,,JEAN ST,,GARDNER,NIVALA SULO W,DUMONT ALBERT W
4168125,01/06/1961,00:00AM,DEED,1,04168/152,,Verified/Certified,,PLEASANT ST,,GARDNER,LANDER RUTH E,CHOJNOWSKI VICTOR J
"
`;
